crates/ty_python_semantic/src/types/infer.rs

```diff
@@ -2134,6 +2134,9 @@ impl<'db, 'ast> TypeInferenceBuilder<'db, 'ast> {
                 if self.in_function_overload_or_abstractmethod() {
                     return;
                 }
+                if self.scope().scope(self.db()).in_type_checking_block() {
+                    return;
+                }
```

Comment:
It's unfortunate that we need to pre-compute this information in semantic index builder considering that empty function bodies in type checking blocks are rare. Ideally, we'd compute it lazily whether the function is inside a `TYPE_CHECKING` block but I don't see how we can do this easily because our AST doesn't allow upward traversal.